File: reactjs/src/Mvc.js
class Mvc {
    static instance = null;

    constructor() {
        this.controllers = {}
        this.models = {};
    }

    static getInstance() {
        if (!Mvc.instance)
            Mvc.instance = new Mvc();
        return Mvc.instance;
    }

    newController(name) {
        var controller = new Controller();
        this.addController(name, controller);
        return controller;
    }

    addController(name, controller) {
        this.controllers[name] = controller;
    }

    getController(name) {
        var controller = this.controllers[name];
        return controller;
    }
}

class Controller {

    constructor() {
        this.views = {};
    }

    addView(action, viewId, view) {
        var available = this.views[action];
        if (!available) {
            available = {};
        }
        available[viewId] = view;
        this.views[action] = available;
    }

    addDefaultView(action, view) {
        this.addView(action, "", view);
    }

    removeView(action, viewId) {
        var available = this.views[action];
        if (available) {
            delete available[viewId];
        }
    }

    removeDefaultView(action) {
        this.removeView(action, "");
    }

    removeViews(action) {
        var available = this.views[action];
        if (available) {
            for (var key in available)
                delete available[key];
        }
        delete this.views[action];
    }

    removeAllViews() {
        for(var action in this.views) {
            var available = this.views[action];
            for (var key in available)
                delete available[key];
            delete this.views[action];
        }
    }

    updateView(action, viewId, data) {
        var available = this.views[action];
        if (available) {
            var view = available[viewId];
            if(view)
                view(data);
        }
    }

    updateViews(action, data) {
        var available = this.views[action];
        if (available) {
            for (var key in available) {
                var view = available[key];
                view(data);
            }
        }
    }
}

export default Mvc

File: android/app/src/main/java/com/tvd12/freechat/Mvc.kt
package com.tvd12.freechat

/**
 * Created by tavandung12 on 10/7/18.
 */

class Mvc private constructor() {
    private val controllers: MutableMap<Any, Any>

    init {
        this.controllers = HashMap()
        this.addController("connection")
        this.addController("contact")
        this.addController("message")
    }

    companion object {
        private val INSTANCE = Mvc()
        fun getInstance() : Mvc = INSTANCE
    }

    private fun addController(name: String) : Controller {
        val controller = Controller()
        addController(name, controller)
        return controller
    }

    private fun addController(name: String, controller: Controller) {
        this.controllers[name] = controller
    }

    fun getController(name: String) : Controller {
        return controllers[name] as Controller
    }
}

class Controller {
    private val views: MutableMap<String, MutableMap<String, IView>>

    init {
        views = HashMap()
    }

    fun addView(action: String, view: IView) {
        addView(action, "", view)
    }

    fun addView(action: String, viewId : String, view: IView) {
        views.computeIfAbsent(action) {
            HashMap()
        }[viewId] = view
    }

    fun removeView(action: String) {
        removeView(action, "")
    }

    fun removeView(action: String, viewId: String) {
        views[action]?.remove(viewId)
    }

    fun updateViews(action: String, data: Any?) {
        views[action]?.keys?.forEach {
            updateView(action, it, data)
        }
    }

    private fun updateView(action: String, viewId: String, data: Any?) {
        val available = views[action]
        if(available != null) {
            available[viewId]?.update(viewId, data)
        }
    }

}

interface IView {

    fun update(viewId: String, data: Any?)
}